File: 652-factorization/1-7-2019.py
class Solution(object):
    def getFactors(self, n):
        """
        :type n: int
        :rtype: List[List[int]]
        """
        result = []
        self.helper(result, [], n, 2)
        return result
    def helper(self, result, item, n, start):
        if n <= 1:
            if len(item) > 1:
                result.append(item[:])
            return
        import math
        for i in xrange(start, int(math.sqrt(n)) + 1):
            if n % i == 0:
                item.append(i)
                self.helper(result, item, n / i, i)
                item.pop()
        if n >= start:
            item.append(n)
            self.helper(result, item, 1, n)
            item.pop()

File: 652-factorization/1-7-2019.java
class Solution {
    public List<List<Integer>> getFactors(int n) {
        if (n < 2) {
            return new ArrayList<>();
        }
        List<List<Integer>> res = new ArrayList<>();
        helper(n, 2, new ArrayList<>(), res);
        return res;
    }
    void helper(int n, int start, List<Integer> list, List<List<Integer>> res) {
        if (n == 1) {
            if (list.size() >= 2) {
                res.add(new ArrayList<>(list));
            }
            return;
        }
        for (int i = start; i <= Math.sqrt(n); i++) {
            if (n % i == 0) {
                list.add(i);
                helper(n / i, i, list, res);
                list.remove(list.size() - 1);
            }
        }
        list.add(n);
        helper(1, n, list, res);
        list.remove(list.size() - 1);
    }
}
